subsys/secure_storage/compatibility/src/its_transform_tsbc.c

```diff
@@ -0,0 +1,194 @@
+	return PSA_SUCCESS;
+}
+
+#if defined(CONFIG_SECURE_STORAGE_ITS_TRANSFORM_TSBC_KEY_PROVIDER_ENTRY_UID_HASH)
+
+static psa_status_t get_key(psa_storage_uid_t uid, uint8_t key[static AEAD_KEY_SIZE])
+{
+	BUILD_ASSERT(AEAD_KEY_SIZE == PSA_HASH_LENGTH(PSA_ALG_SHA_256));
+	size_t hash_length;
+
+	return psa_hash_compute(PSA_ALG_SHA_256, (uint8_t *)&uid, sizeof(uid), key, AEAD_KEY_SIZE,
+				&hash_length);
+}
+
+#elif defined(CONFIG_SECURE_STORAGE_ITS_TRANSFORM_TSBC_KEY_PROVIDER_HUK_LIBRARY)
+#include <hw_unique_key.h>
+
+static psa_status_t get_key(psa_storage_uid_t uid, uint8_t key[static AEAD_KEY_SIZE])
+{
+	int result;
+	enum hw_unique_key_slot key_slot;
+
+#ifdef HUK_HAS_KMU
+	key_slot = HUK_KEYSLOT_MKEK;
+#else
+	key_slot = HUK_KEYSLOT_KDR;
+#endif
+
+	result = hw_unique_key_derive_key(key_slot, NULL, 0, (uint8_t *)&uid, sizeof(uid), key,
+					  AEAD_KEY_SIZE);
+	if (result != HW_UNIQUE_KEY_SUCCESS) {
+		return PSA_ERROR_BAD_STATE;
+	}
+
+	return PSA_SUCCESS;
+}
+
+#endif /* CONFIG_SECURE_STORAGE_ITS_TRANSFORM_TSBC_KEY_PROVIDER */
+
+struct stored_entry_header {
+	psa_storage_create_flags_t create_flags;
+	size_t data_len;
+};
+
+struct stored_entry {
+	struct stored_entry_header header;
+	uint8_t nonce[AEAD_NONCE_SIZE];
+	uint8_t ciphertext[CIPHERTEXT_MAX_SIZE];
+} __packed;
+BUILD_ASSERT(sizeof(struct stored_entry) == SECURE_STORAGE_ITS_TRANSFORM_MAX_STORED_DATA_SIZE);
+
+/** @return The length of a `struct stored_entry` whose `ciphertext` is `len` bytes long. */
+#define STORED_ENTRY_LEN(len) (sizeof(struct stored_entry) - CIPHERTEXT_MAX_SIZE + len)
+
+static psa_status_t crypt(psa_key_usage_t operation, psa_storage_uid_t uid,
+			  const struct stored_entry *stored_entry,
+			  size_t input_len, const void *input,
+			  size_t output_size, void *output, size_t *output_len)
+{
+	psa_status_t ret;
+	psa_key_attributes_t key_attributes = PSA_KEY_ATTRIBUTES_INIT;
+	uint8_t key[AEAD_KEY_SIZE];
+	psa_status_t (*aead_crypt)(const psa_key_attributes_t *attributes, const uint8_t *key,
+				   size_t key_size, psa_algorithm_t alg, const uint8_t *nonce,
+				   size_t nonce_length, const uint8_t *add_data,
+				   size_t add_data_len, const uint8_t *input, size_t input_len,
+				   uint8_t *output, size_t output_size, size_t *output_len);
+
+	psa_set_key_usage_flags(&key_attributes, operation);
+	psa_set_key_lifetime(&key_attributes, PSA_KEY_LIFETIME_VOLATILE);
+	psa_set_key_type(&key_attributes, PSA_KEY_TYPE_CHACHA20);
+	psa_set_key_algorithm(&key_attributes, PSA_ALG_CHACHA20_POLY1305);
+	psa_set_key_bits(&key_attributes, PSA_BYTES_TO_BITS(AEAD_KEY_SIZE));
+
+	ret = get_key(uid, key);
+	if (ret != PSA_SUCCESS) {
+		return ret;
+	}
+
+	/* Avoid calling psa_aead_*crypt() because that would require importing keys into
+	 * PSA Crypto. This gets called from PSA Crypto for storing persistent keys so,
+	 * even if using PSA_KEY_LIFETIME_VOLATILE, it would corrupt the global key store
+	 * which holds all the active keys in the PSA Crypto core.
```

Comment:
Is this about Oberon?
```suggestion
	 * which holds all the active keys in the Oberon PSA Core.
```

Reply:
No.